File: lib/OUGP/Population.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

namespace OneUnified.GeneticProgramming {
  class Population {

    // page 114 Koza 1992
    public int PopulationSize = 20; // should be some multiple of ( 2 * ( size / maxdepth - 1 ) )
    public int MaxGenerations = 25;

    public double probCrossover = 0.95;
    public double probReproduction = 0.10;

    public double probFunctionPointCrossover = 0.90;
    public double probTerminalPointCrossover = 0.10;

    public int MaxDepthOnCreation = 4;
    public int MaxDepthOnCrossover = 17;

    public double probMutation = 0.0;
    public double probPermutation = 0.0;

    public int TournamentSize = 2;

    public double probDecimation = 0.58;
    public double ratioElitism = 0.012;

    public int cntSignals = 4; // LongEnter, LongExit, ShortEnter, ShortExit

    //public Individual[] individuals;
    public Individual[] curGeneration;
    private Individual[] nxtGeneration;
    private ArrayList Generations;
    public int cntGenerations = 0;

    public int cntAboveAvg = 0;

    private Random random;

    public Population( int PopulationSize ) {

      // initialize sizes and random number generator
      random = new Random();
      this.PopulationSize = PopulationSize;
      Generations = new ArrayList(MaxGenerations);

      // allocate a generation of individuals
      curGeneration = new Individual[PopulationSize];
      cntGenerations = 1;
      Generations.Add(curGeneration);

      CreateIndividuals(PopulationSize, curGeneration);

    }

    private void CreateIndividuals( int Count, Individual[] Generation ) {

      // can only populate starting at beginning

      // create individuals with RootNode signals
      // keep rootnodes in a list so they can be populated with random expressions
      ArrayList rn = new ArrayList(cntSignals * Count);
      for (int i = 0; i < Count; i++) {
        Individual individual = new Individual();
        Generation[i] = individual;
        rn.Add(individual.LongEnter);
        rn.Add(individual.LongExit);
        rn.Add(individual.ShortEnter);
        rn.Add(individual.ShortExit);
      }

      // implement ramped half and half, koza, 1992, page 93
      RootNode node;
      int nodesperstep = rn.Count / (MaxDepthOnCreation - 1);
      int nodesperhalf = nodesperstep / 2;
      for (int step = 2; step <= MaxDepthOnCreation; step++) {
        for (int i = 1; i <= nodesperhalf; i++) {
          int r = random.Next(rn.Count - 1);
          node = rn[r] as RootNode;
          rn[r] = null;
          rn.RemoveAt(r);
          node.BuildTree(true, true, step); // random tree

          r = random.Next(rn.Count - 1);
          node = rn[r] as RootNode;
          rn[r] = null;
          rn.RemoveAt(r);
          node.BuildTree(false, true, step); // full tree
        }
      }

      // due to rounding above, finish up any unused rootnodes
      while (0 != rn.Count) {
        node = rn[0] as RootNode;
        rn[0] = null;
        rn.RemoveAt(0);
        node.BuildTree(true, true, MaxDepthOnCreation);  // random tree
      }
    }

    private int TournamentSelection() {
      double prob1; // use above average individuals % of the time
      int prob2;  // used for selecting individual
      int result = PopulationSize - 1; // default to worst individual (array is sorted)
      for (int i = TournamentSize; i >= 1; i--) {
        prob1 = random.NextDouble();
        if (prob1 < 0.35) {
          prob2 = random.Next(cntAboveAvg - 1);
          result = Math.Min(result, prob2);
        }
        else {
          prob2 = random.Next(cntAboveAvg, PopulationSize - 1);
          result = Math.Min(result, prob2);
        }
      }
      return result;
    }

    private bool CheckForMatchInGeneration( Individual individual, Individual[] Generation, int MaxIndex ) {
      bool b = false;
      for (int i = 0; i <= MaxIndex; i++) {
        if (individual.ToString() == Generation[i].ToString()) {
          b = true;
          break;
        }
      }
      return b;
    }

    public bool MakeNewGeneration( bool CopyValues ) {
      cntGenerations++;
      bool bMore = false;
      if (MaxGenerations >= cntGenerations) {

        bMore = true;
        nxtGeneration = new Individual[PopulationSize];
        Generations.Add(nxtGeneration);

        int ix = 0;

        // decimate and repopulate everything at average and below
        int decimation = (int)Math.Min(0.40 * PopulationSize, PopulationSize - cntAboveAvg);
        // always introduce a certain minimum set of new individuals
        decimation = Math.Max(decimation, (int)(0.04 * PopulationSize));
        Console.WriteLine("Decimating {0} individuals", decimation);
        CreateIndividuals(decimation, nxtGeneration);
        ix += decimation;

        // elitism
        int j = 0;
        int cntElites = (int)(ratioElitism * PopulationSize);
        int cntAcceptedElites = 0;
        if (0 == ix && cntElites>0) {
          nxtGeneration[ix] = curGeneration[j];
          ix++;
          j++;
          cntAcceptedElites++;
        }
        while ( cntAcceptedElites < cntElites ) {
          if (!CheckForMatchInGeneration(curGeneration[j],nxtGeneration,ix-1)) {
            nxtGeneration[ix] = curGeneration[j];
            ix++;
            cntAcceptedElites++;
          }
          j++;
        }

        // reproduction & crossover
        int prob;
        while (ix < PopulationSize) {
        //for (int i = PopulationSize - ix; ix >= 1; i--) {
          double prob1 = random.NextDouble();
          if (prob1 < probReproduction || ( PopulationSize - ix ) == 1 ) {
            // reproduction
            //prob = random.Next(curGeneration.Length - 1);
            nxtGeneration[ix++] = curGeneration[TournamentSelection()];
          }
          else {
            // crossover
            bool bOk = true;
            Individual i1;
            Individual i2;
            //prob = random.Next(curGeneration.Length - 1);
            i1 = curGeneration[TournamentSelection()];
            //prob = random.Next(curGeneration.Length - 1);
            i2 = curGeneration[TournamentSelection()];

            RootNode newxvr1; // seed and grow crossovers here
            RootNode newxvr2;
            RootNode oldxvr1; // obtain seed from here
            RootNode oldxvr2;
            RootNode node;  // holds node

            // load array with references to old nodes
            ArrayList nodesSrc = new ArrayList(2 * cntSignals);

            nodesSrc.Add(i1.LongEnter);
            nodesSrc.Add(i1.LongExit);
            nodesSrc.Add(i1.ShortEnter);
            nodesSrc.Add(i1.ShortExit);
            nodesSrc.Add(i2.LongEnter);
            nodesSrc.Add(i2.LongExit);
            nodesSrc.Add(i2.ShortEnter);
            nodesSrc.Add(i2.ShortExit);

            // create new individuals and provide references to new nodes
            i1 = new Individual();
            i2 = new Individual();

            ArrayList nodesDst = new ArrayList(2 * cntSignals);

            nodesDst.Add(i1.LongEnter);
            nodesDst.Add(i1.LongExit);
            nodesDst.Add(i1.ShortEnter);
            nodesDst.Add(i1.ShortExit);
            nodesDst.Add(i2.LongEnter);
            nodesDst.Add(i2.LongExit);
            nodesDst.Add(i2.ShortEnter);
            nodesDst.Add(i2.ShortExit);

            // crossover all but two random pairs
            for (int k = cntSignals - 1; k >= 1; k-- ) {

              prob = random.Next(nodesSrc.Count - 1);
              oldxvr1 = nodesSrc[prob] as RootNode;
              nodesSrc.RemoveAt(prob);

              prob = random.Next(nodesSrc.Count - 1);
              oldxvr2 = nodesSrc[prob] as RootNode;
              nodesSrc.RemoveAt(prob);

              newxvr1 = oldxvr1.Replicate(CopyValues) as RootNode;
              newxvr2 = oldxvr2.Replicate(CopyValues) as RootNode;

              newxvr1.RebuildCandidateLists();
              newxvr2.RebuildCandidateLists();

              bOk = bOk && CrossOver(newxvr1, newxvr2);

              node = nodesDst[0] as RootNode;
              nodesDst.RemoveAt(0);
              node = newxvr1;
              node = nodesDst[0] as RootNode;
              nodesDst.RemoveAt(0);
              node = newxvr2;

            }

            // crossover last two pairs
            oldxvr1 = nodesSrc[0] as RootNode;
            nodesSrc.RemoveAt(0);
            oldxvr2 = nodesSrc[0] as RootNode;
            nodesSrc.RemoveAt(0);

            newxvr1 = oldxvr1.Replicate(CopyValues) as RootNode;
            newxvr2 = oldxvr2.Replicate(CopyValues) as RootNode;

            newxvr1.RebuildCandidateLists();
            newxvr2.RebuildCandidateLists();

            bOk = bOk && CrossOver(newxvr1, newxvr2);

            node = nodesDst[0] as RootNode;
            nodesDst.RemoveAt(0);
            node = newxvr1;
            node = nodesDst[0] as RootNode;
            nodesDst.RemoveAt(0);
            node = newxvr2;

            if (bOk) {
              nxtGeneration[ix++] = i1;
              nxtGeneration[ix++] = i2;
            }

          }
        }
        curGeneration = nxtGeneration;
      }
      return bMore;
    }

    private bool CrossOver( RootNode rn1, RootNode rn2 ) {

      bool bSuccessful = false;
      int prob;
      Node node1 = null;
      bool b = false;
      bool d = false;

      if ( rn1.alBoolCandidateNodes.Count > 0
        && rn2.alBoolCandidateNodes.Count > 0
        && rn1.alDblCandidateNodes.Count > 0
        && rn2.alDblCandidateNodes.Count > 0 
        ) {
        prob = random.Next(rn1.alAllCandidateNodes.Count - 1);
        node1 = rn1.alAllCandidateNodes[prob] as Node;
        b = node1 is BoolNode;
        d = node1 is DoubleNode;
      }
      else {
        if ( rn1.alBoolCandidateNodes.Count > 0
          && rn2.alBoolCandidateNodes.Count > 0 
          ) {
          prob = random.Next(rn1.alBoolCandidateNodes.Count - 1);
          node1 = rn1.alBoolCandidateNodes[prob] as Node;
          b = node1 is BoolNode;
        }
        else {
          if ( rn1.alDblCandidateNodes.Count > 0
            && rn2.alDblCandidateNodes.Count > 0 
            ) {
            prob = random.Next(rn1.alDblCandidateNodes.Count - 1);
            node1 = rn1.alDblCandidateNodes[prob] as Node;
            d = node1 is DoubleNode;
          }
        }
      }

      if (b || d) {
        Node node2 = null;
        if (b) {
          prob = random.Next(rn2.alBoolCandidateNodes.Count - 1);
          node2 = rn2.alBoolCandidateNodes[prob] as Node;
        }
        if (d) {
          prob = random.Next(rn2.alDblCandidateNodes.Count - 1);
          node2 = rn2.alDblCandidateNodes[prob] as Node;
        }

        Node Parent1 = node1.Parent;
        Node Parent2 = node2.Parent;

        switch (node1.ParentSide) {
          case Node.EParentLink.Left:
            Parent1.AddLeft(node2);
            break;
          case Node.EParentLink.Center:
            Parent1.AddCenter(node2);
            break;
          case Node.EParentLink.Right:
            Parent1.AddRight(node2);
            break;
        }

        switch (node2.ParentSide) {
          case Node.EParentLink.Left:
            Parent2.AddLeft(node1);
            break;
          case Node.EParentLink.Center:
            Parent2.AddCenter(node1);
            break;
          case Node.EParentLink.Right:
            Parent2.AddRight(node1);
            break;
        }
        bSuccessful = true;
      }
      return bSuccessful;
    }

    public void CalcFitness() {

      double max = 0;
      double min = 0;
      int cntIndividuals = 0;
      foreach (Individual i in curGeneration) {
        if (0 == cntIndividuals) {
          min = i.RawFitness;
          max = i.RawFitness;
        }
        else {
          max = Math.Max(max, i.RawFitness);
          min = Math.Min(min, i.RawFitness);
        }
        cntIndividuals++;
      }
      double mid = (max - min) / 2.0;

      double sum = 0;
      foreach (Individual i in curGeneration) {
        i.RelativeFitness = max - i.RawFitness;
        i.AdjustedFitness = 1.0 / (1.0 + i.RelativeFitness);
        sum += i.AdjustedFitness;
      }

      cntAboveAvg = 0;
      double mean = sum / cntIndividuals;
      foreach (Individual i in curGeneration) {
        i.NormalizedFitness = i.AdjustedFitness / sum;
        cntAboveAvg += (i.AdjustedFitness > mean) ? 1 : 0;
      }

      Console.WriteLine("cntAboveAvg = {0}", cntAboveAvg);

      IndividualComparison CompareIndividuals = new IndividualComparison();
      Array.Sort(curGeneration, CompareIndividuals);

    }
  }
}

File: lib/OUGP/Population.cpp
/************************************************************************
 * Copyright(c) 2012, One Unified. All rights reserved.                 *
 * email: info@oneunified.net                                           *
 *                                                                      *
 * This file is provided as is WITHOUT ANY WARRANTY                     *
 *  without even the implied warranty of                                *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.                *
 *                                                                      *
 * This software may not be used nor distributed without proper license *
 * agreement.                                                           *
 *                                                                      *
 * See the file LICENSE.txt for redistribution information.             *
 ************************************************************************/

#include <vector>
#include <algorithm>

#include <boost/phoenix/core.hpp>
#include <boost/phoenix/operator.hpp>
#include <boost/phoenix/core/reference.hpp>
#include <boost/phoenix/stl/container.hpp>

#include "Individual.h"

#include "Population.h"

namespace ou { // One Unified
namespace gp { // genetic programming

// assumption:  each population and individual will be sharing the same set of Node types.  Additional non default Node types should only be 
//  registered once

Population::Population( unsigned int nPopulationSize ) 
  : m_nPopulationSize( nPopulationSize ), m_dblPopulationSize( nPopulationSize ),
  m_nMaxGenerations( 25 ), m_nMaxDepthOnCreation( 6 ), m_nMaxDepthOnCrossover( 17 ),
  m_probCrossover( 0.95 ), m_probReproduction( 0.10 ), m_probFunctionPointCrossover( 0.90 ), m_probTerminalPointCrossover( 0.10), 
  m_probMutation( 0.0 ), m_probPermutation( 0.0 ), m_probDecimation( 0.58 ), m_ratioElitism( 0.012 ),
  m_probTournamentSegregation( 0.35 ),
  m_nTournamentSize( 2 ), 
  m_rng( std::time( 0 ) ),  // possible issue after jan 18, 2038?
  m_urd( 0.0, 1.0 ),  // probability in [0.0, 1.0)
  m_cntAboveAverage( 0 )
{
}

Population::~Population(void) {
  for ( vGenerations_t::iterator iter = m_vGenerations.begin(); m_vGenerations.end() != iter; ++iter ) {
    delete *iter;
  }
}

void Population::BuildIndividuals( vGeneration_t& vGeneration ) {

  using boost::phoenix::arg_names::arg1;

  // can only populate starting at beginning
  // create individuals with RootNode signals
  // keep rootnodes in a list so they can be populated with random expressions
  typedef Individual::pRootNode_t pRootNode_t;
  typedef std::vector<pRootNode_t*> vSignals_t;
  vSignals_t vSignals;
  unsigned int n( vGeneration.size() * Individual::Signals_t::cntSignals );
  vSignals.reserve( n );
  for ( vSignals_t::size_type ix = 0; ix < vGeneration.size(); ++ix ) {
    vGeneration[ ix ].m_Signals.EachSignal( boost::phoenix::push_back( boost::phoenix::ref(vSignals), arg1 ) );
  }

  // implement ramped half and half, koza, 1992, page 93
  pRootNode_t* ppRootNode;
  unsigned int nNodesPerStep = n / ( m_nMaxDepthOnCreation - 1 );
  unsigned int nNodesPerHalf = nNodesPerStep / 2;

  vSignals_t::const_iterator iter;
  vSignals_t::size_type ix;
  
  assert( n >= ( ( m_nMaxDepthOnCreation - 1 ) * nNodesPerHalf * 2 ) );
  for ( unsigned int step1 = 2; step1 <= m_nMaxDepthOnCreation; ++step1 ) {
    for ( unsigned int step2 = 1; step2 <= nNodesPerHalf; ++step2 ) {

      boost::random::uniform_int_distribution<vSignals_t::size_type> dist1( 0, vSignals.size() - 1 );  // closed range
      ix = dist1( m_rng );
      iter = vSignals.begin() + ix;
      ppRootNode = *iter;
      vSignals.erase( iter );
      assert( 0 != ppRootNode );
      assert( 0 == *ppRootNode );
      *ppRootNode = new RootNode;
      m_tb.BuildTree( **ppRootNode, true, true, step1 );  // random tree

      boost::random::uniform_int_distribution<vSignals_t::size_type> dist2( 0, vSignals.size() - 1 );  // closed range
      ix = dist2( m_rng );
      iter = vSignals.begin() + ix;
      ppRootNode = *iter;
      vSignals.erase( iter );
      assert( 0 != ppRootNode );
      assert( 0 == *ppRootNode );
      *ppRootNode = new RootNode;
      m_tb.BuildTree( **ppRootNode, false, true, step1 );  // full tree
    }
  }
  // due to rounding above, finish up any unused rootnodes
  while ( 0 != vSignals.size() ) {
    iter = vSignals.begin();  // change this to use end of vector instead
    ppRootNode = *iter;
    vSignals.erase( iter );
    assert( 0 != ppRootNode );
    assert( 0 == *ppRootNode );
    *ppRootNode = new RootNode;
    m_tb.BuildTree( **ppRootNode, true, true, m_nMaxDepthOnCreation );  // full tree
  }
}

unsigned int Population::TournamentSelection( unsigned int cntAboveAverage ) {
  double prob1; // use above average individuals % of the time
  unsigned int prob2; // used for selecting individual
  unsigned int result = m_nPopulationSize - 1;  // default to worst individual (vector is sorted)
  boost::random::uniform_int_distribution<unsigned int> dist1( 0, cntAboveAverage - 1 );
  boost::random::uniform_int_distribution<unsigned int> dist2( cntAboveAverage, m_nPopulationSize - 1 );
  for ( unsigned int i = m_nTournamentSize; i>=1; --i ) {
    prob1 = m_urd( m_rng );
    if ( m_probTournamentSegregation > prob1 ) {
      prob2 = dist1( m_rng );
    }
    else {
      prob2 = dist2( m_rng );
    }
    result = std::min<unsigned int>( result, prob2 );
  }
  return result;
}

bool Population::IsMatchInGeneration( const Individual& individual, const vGeneration_t& generation, vGeneration_t::size_type ixMax ) {
  bool b = false;
  std::stringstream ssIndividual1;
  individual.TreeToString( ssIndividual1 );
  for ( vGenerations_t::size_type ix = 0; ix <= ixMax; ++ix ) {
    std::stringstream ssIndividual2;
    generation[ ix ].TreeToString( ssIndividual2 );
    if ( ssIndividual1.str() == ssIndividual2.str() ) {
      b = true;
      break;
    }
  }
  return b;
}

bool Population::MakeNewGeneration( bool bCopyValues ) {

  using boost::phoenix::arg_names::arg1;

  bool bMore( false );
  unsigned int cntMaxElites( (unsigned int) std::floor( m_ratioElitism * m_dblPopulationSize + 0.5 ) );

  if ( 0 == m_vGenerations.size() ) {
    bMore = true;
    m_pvCurGeneration = new vGeneration_t;
    m_pvCurGeneration->resize( m_nPopulationSize );
    BuildIndividuals( *m_pvCurGeneration );
    m_vGenerations.push_back( m_pvCurGeneration );
  }
  else if ( m_nMaxGenerations > m_vGenerations.size() + 1 ) { // is this a correct validation?

    bMore = true;
    m_pvNxtGeneration = new vGeneration_t;  // set no individuals

//    assert( 0 != m_cntAboveAverage );  // put back in sometime

    double dblPopulationSize( m_nPopulationSize );
    // decimate and repopulate minimum of ( 40% population, count below average )
    unsigned int decimation = std::min<unsigned int>( (unsigned int) std::floor( 0.40 * m_dblPopulationSize + 0.5 ), m_nPopulationSize - m_cntAboveAverage );
    // always introduce a certain minimum set of new individuals
    decimation = std::max<unsigned int>( decimation, (unsigned int) std::floor( 0.04 * m_dblPopulationSize + 0.5 ) );

    m_pvNxtGeneration->resize( decimation );
    BuildIndividuals( *m_pvNxtGeneration );  // build the beginning population
    m_vGenerations.push_back( m_pvNxtGeneration );

    // elitism
    unsigned int cntAcceptedElites( 0 );
    vGeneration_t::iterator iter( m_pvCurGeneration->begin() );
    if ( ( 0 == m_pvNxtGeneration->size() ) && ( 0 < cntMaxElites ) ) { // bring over at least one elite
      m_pvNxtGeneration->push_back( *iter );
      ++cntAcceptedElites;
    }
    while ( cntAcceptedElites < cntMaxElites ) {
      iter++;
      if ( m_pvCurGeneration->end() == iter ) break;
      if ( !IsMatchInGeneration( *iter, *m_pvNxtGeneration, m_pvNxtGeneration->size() - 1 ) ) {
        m_pvNxtGeneration->push_back( *iter );
        ++cntAcceptedElites;
      }
    }

    // reproduction and crossover
    while ( m_nPopulationSize > m_pvNxtGeneration->size() ) {
      double prob1( m_urd( m_rng ) );
      if ( ( m_probReproduction > prob1 ) || ( 1 == ( m_nPopulationSize - m_pvNxtGeneration->size() ) ) ) {  // cross over appends two individuals
        // reproduction
        const Individual& test( m_pvCurGeneration->at( TournamentSelection( m_cntAboveAverage ) ) );
        if ( !IsMatchInGeneration( test, *m_pvNxtGeneration, m_pvNxtGeneration->size() - 1 ) ) {
          m_pvNxtGeneration->push_back( test ); 
        }
      }
      else {
        // cross over
        bool bOk( true );

        Individual& indvlSrc1( m_pvCurGeneration->at( TournamentSelection( m_cntAboveAverage ) ) );
        Individual& indvlSrc2( m_pvCurGeneration->at( TournamentSelection( m_cntAboveAverage ) ) );

        pRootNode_t rnNewXOver1; // seed and grow crossovers here
        pRootNode_t rnNewXOver2;
        pRootNode_t rnOldXOver1; // obtain seed from here
        pRootNode_t rnOldXOver2;
        pRootNode_t rnNode( new RootNode() );  // holds node

        std::vector<pRootNode_t*> vSrcNodes;
        indvlSrc1.m_Signals.EachSignal( push_back( boost::phoenix::ref( vSrcNodes ), arg1 ) );
        indvlSrc2.m_Signals.EachSignal( push_back( boost::phoenix::ref( vSrcNodes ), arg1 ) );

        Individual indvlDst1;
        Individual indvlDst2;

        std::vector<pRootNode_t*> vDstNodes;

        indvlDst1.m_Signals.EachSignal( push_back( boost::phoenix::ref( vDstNodes ), arg1 ) );
        indvlDst2.m_Signals.EachSignal( push_back( boost::phoenix::ref( vDstNodes ), arg1 ) );

        // crossover all but last random pair
        typedef std::vector<pRootNode_t>::size_type prob2_t;
        prob2_t prob2;
        for ( unsigned int k = Individual::Signals_t::cntSignals - 1; k >= 1; --k ) {
          boost::random::uniform_int_distribution<prob2_t> dist1( 0, vSrcNodes.size() - 1 );  // closed range
          prob2 = dist1( m_rng );
          rnOldXOver1 = *vSrcNodes[ prob2 ];
          vSrcNodes.erase( vSrcNodes.begin() + prob2 );

          boost::random::uniform_int_distribution<prob2_t> dist2( 0, vSrcNodes.size() - 1 );  // closed range
          prob2 = dist2( m_rng );
          rnOldXOver2 = *vSrcNodes[ prob2 ];
          vSrcNodes.erase( vSrcNodes.begin() + prob2 );

          rnNewXOver1 = dynamic_cast<RootNode*>( rnOldXOver1->Replicate( bCopyValues ) );
          rnNewXOver2 = dynamic_cast<RootNode*>( rnOldXOver2->Replicate( bCopyValues ) );

          // rebuild candidate lists here, if we need them
          rnNewXOver1->PopulateCandidates( &m_rng );
          rnNewXOver2->PopulateCandidates( &m_rng );

          // do the crossover
          CrossOver( rnNewXOver1, rnNewXOver2 );

          *vDstNodes.back() = rnNewXOver1;
          vDstNodes.pop_back();
          *vDstNodes.back() = rnNewXOver2;
          vDstNodes.pop_back();
        }

        // cross over last pair
        rnOldXOver1 = *vSrcNodes.back();
        vSrcNodes.pop_back();
        rnOldXOver2 = *vSrcNodes.back();
        vSrcNodes.pop_back();
        assert( 0 == vSrcNodes.size() );

        rnNewXOver1 = dynamic_cast<RootNode*>( rnOldXOver1->Replicate( bCopyValues ) );
        rnNewXOver2 = dynamic_cast<RootNode*>( rnOldXOver2->Replicate( bCopyValues ) );

        rnNewXOver1->PopulateCandidates( &m_rng );
        rnNewXOver2->PopulateCandidates( &m_rng );

        CrossOver( rnNewXOver1, rnNewXOver2 );

        *vDstNodes.back() = rnNewXOver1;
        vDstNodes.pop_back();
        *vDstNodes.back() = rnNewXOver2;
        vDstNodes.pop_back();

        // add to generation
        m_pvNxtGeneration->push_back( indvlDst1 ); // does this properly copy Signals_t?
        m_pvNxtGeneration->push_back( indvlDst2 ); 
      }
      m_pvCurGeneration = m_pvNxtGeneration;
    }
  }

  return bMore;
}

bool Population::CrossOver( pRootNode_t& rn1, pRootNode_t& rn2 ) {

  bool bSuccessful = true;

  Node* node1( 0 );

  NodeType::E ntCrossOver;

  assert( rn1->HasBooleanCandidates() );
  assert( rn2->HasBooleanCandidates() );

  if ( !rn1->HasDoubleCandidates() || !rn2->HasDoubleCandidates() ) {
    // crossover on boolean candidates, there will always be boolean candidates if one or the other doesn't have double candidates
    node1 = rn1->RandomBooleanCandidate();
    assert( NodeType::Bool == node1->ReturnType() );
    ntCrossOver = NodeType::Bool;
  }
  else {
    // crossover on all candidates
    node1 = rn1->RandomAllCandidate();
    ntCrossOver = node1->ReturnType();
  }

  Node* node2( 0 );
  switch ( ntCrossOver ) {
  case NodeType::Bool:
    node2 = rn2->RandomBooleanCandidate();
    break;
  case NodeType::Double:
    node2 = rn2->RandomDoubleCandidate();
    break;
  }

  Node& parent1( node1->Parent() );
  Node& parent2( node2->Parent() );

  switch ( node1->ParentSide() ) {
  case ParentLink::Left:
    parent1.AddLeft( node2 );
    break;
  case ParentLink::Center:
    parent1.AddCenter( node2 );
    break;
  case ParentLink::Right:
    parent1.AddRight( node2 );
    break;
  }

  switch ( node2->ParentSide() ) {
  case ParentLink::Left:
    parent2.AddLeft( node1 );
    break;
  case ParentLink::Center:
    parent2.AddCenter( node1 );
    break;
  case ParentLink::Right:
    parent2.AddRight( node1 );
    break;
  }

  return bSuccessful;
}

void Population::CalcFitness( void ) {
  double dblMax( 0.0 );
  double dblMin( 0.0 );

  for ( vGeneration_t::iterator iter = (*m_pvCurGeneration).begin(); (*m_pvCurGeneration).end() != iter; ++iter ) {
    if ( (*m_pvCurGeneration).begin() == iter ) {
      dblMin = iter->m_dblRawFitness;
      dblMax = iter->m_dblRawFitness;
    }
    else {
      dblMin = std::min<double>( dblMin, iter->m_dblRawFitness );
      dblMax = std::max<double>( dblMax, iter->m_dblRawFitness );
    }
  }

  double dblSum( 0.0 );
  for ( vGeneration_t::iterator iter = (*m_pvCurGeneration).begin(); (*m_pvCurGeneration).end() != iter; ++iter ) {
    iter->m_dblRelativeFitness = dblMax - iter->m_dblRawFitness;
    dblSum += iter->m_dblAdjustedFitness = 1.0 / ( 1.0 + iter->m_dblRelativeFitness );
  }

  m_cntAboveAverage = 0;
  double dblMean = dblSum / m_pvCurGeneration->size();
  for ( vGeneration_t::iterator iter = (*m_pvCurGeneration).begin(); (*m_pvCurGeneration).end() != iter; ++iter ) {
    iter->m_dblNormalizedFitness = iter->m_dblAdjustedFitness / dblSum;
    m_cntAboveAverage += ( iter->m_dblAdjustedFitness > dblMean ) ? 1 : 0;
  }

  using boost::phoenix::arg_names::arg1;
  using boost::phoenix::arg_names::arg2;

  vGeneration_t& gen( *m_pvCurGeneration );

  std::sort( gen.begin(), gen.end(), arg1 > arg2 );
}

} // namespace gp
} // namespace ou
